### ISSUE TITLE
Unqualified references in uncorrelated subqueries

### ISSUE THREAD

**Issue by mroy-seedbox** [sqlfluff/sqlfluff#6326]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

When using simple _uncorrelated_ subqueries in a `WHERE` clause (in order to filter records), [RF02](https://docs.sqlfluff.com/en/stable/reference/rules.html#rule-RF02) is triggered.

### Use case

When a subquery is uncorrelated (especially if it does not even contain a `WHERE` clause), its content shouldn't need to be qualified.

For example:
```sql
SELECT *
FROM tbl
WHERE val = (SELECT MAX(val) FROM tbl);
```

I don't see what would be the value in qualifying the subquery:
```sql
SELECT *
FROM tbl
WHERE val = (SELECT MAX(tbl.val) FROM tbl);
```

It's almost more confusing, especially when it is the same table being used twice.

### Dialect

Snowflake (but should apply to any dialect).

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

I'm not sure where to start. I think this one might be beyond me.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by rogalski** [comment]
regression: 92e1bf376
CC @keraion

IMO check needs a mode switch, _standard_ which will treat subqueries as boundaries, and _strict_ which would complain (possibly with customized error message).

**Comment by mroy-seedbox** [comment]
Actually, @keraion 's rationale isn't bad! With a good rationale like that, we're willing to adapt. 👍 But I don't know about everyone else. 🤷‍♂️ 

Although it's still possible to have that same ambiguity if we reuse aliases inside of subqueries:
```sql
SELECT *
FROM tbl AS t
WHERE t.val IN (SELECT t.val FROM tbl2 AS t)
```

Or if we use a conflicting alias:
```sql
SELECT *
FROM tbl
WHERE val IN (SELECT tbl.val FROM tbl2 AS tbl)
```

Or from the example above:
```sql
SELECT *
FROM tbl
WHERE val = (SELECT MAX(tbl.val) FROM tbl);
```

All of which are currently valid as of SQLFluff `3.2.3`.

So in strict mode, it might be good if SQLFluff made sure that aliases are only used once (including those inside of subqueries). Although this has some overlap with [AL04](https://docs.sqlfluff.com/en/stable/reference/rules.html#rule-AL04), so maybe we would need a strict mode there as well...

Or instead of `standard` vs `strict` for `AL04`, maybe `local` vs `global` would be a bit more explicit:
- `local` (the default/current behavior): "Table aliases should be unique within each clause."
- `global`: "Table aliases should be globally unique within a query."
  - Although CTEs are segregated from each other (so no danger/ambiguity there is the same alias is reused across CTEs), so perhaps `global` is not the right word.
  - So maybe instead of `local` vs `global`, what we would need is an `include_subqueries` boolean toggle.

And then the documentation for `RF02` should probably link to `AL04` to explain this case. Unless we make `include_subqueries = true` the default.

And then the ambiguity I posted above would also disappear, because the `tbl` in the subquery would need to be aliased (to something new/unused) in order to avoid triggering `AL04`:
```sql
SELECT *
FROM tbl
WHERE val = (SELECT MAX(t.val) FROM tbl AS t);
```

**Comment by keraion** [comment]
Thank you both for your comments! I'm in agreement with what @mroy-seedbox has listed out and agree that this has an impact which other rules should be responsible for. I'll try to pull something together with this information.
